src/ObjCRuntime/Class.cs

```diff
@@ -482,55 +482,58 @@ static Module ResolveModule (Assembly assembly, uint token)
 
 // Restrict this code to desktop for now, which is where most of the problems with outdated generated static registrar code occur.
 #if __MACOS__ || __MACCATALYST__
-		static bool verified_static_registrar_code;
-		unsafe static void VerifyStaticRegistrarCode ()
+		static bool? verify_static_registrar_code;
+		static object? verification_lock;
+		static Dictionary<IntPtr, object?>? verified_assemblies; // Use Dictionary instead of HashSet to avoid pulling in System.Core.dll.
+		unsafe static void VerifyStaticRegistrarCode (IntPtr assembly_name, Assembly assembly)
 		{
-			if (verified_static_registrar_code)
+			if (verify_static_registrar_code is null) {
+				verify_static_registrar_code = !string.IsNullOrEmpty (Environment.GetEnvironmentVariable ("XAMARIN_VALIDATE_STATIC_REGISTRAR_CODE"));
+				verification_lock = new object ();
+			}
+			if (verify_static_registrar_code != true)
 				return;
-			verified_static_registrar_code = true;
 
-			if (!string.IsNullOrEmpty (Environment.GetEnvironmentVariable ("XAMARIN_SKIP_STATIC_REGISTRAR_CODE_VALIDATION")))
-				return;
+			lock (verification_lock!) {
+				if (verified_assemblies is null) {
+					verified_assemblies = new Dictionary<IntPtr, object?> (Runtime.IntPtrEqualityComparer);
+				} else if (verified_assemblies.ContainsKey (assembly_name)) {
+					return;
+				}
+				verified_assemblies [assembly_name] = null;
+			}
 
 			var map = Runtime.options->RegistrationMap;
-			List<Exception>? exceptions = null;
-
 			if (map is null)
 				return;
 
 			for (var i = 0; i < map->assembly_count; i++) {
 				var entry = map->assemblies [i];
 				var name = Marshal.PtrToStringAuto (entry.name)!;
+				if (!Runtime.StringEquals (assembly_name, name))
```

Comment:
I'm confused here a bit, why not just compare the two byte sequences?

Reply:
Don't need to marshal the string to do the comparison, I think?

Reply:
Because I couldn't think of any built-in method to do that, and I didn't feel like writing a new implementation either :)

Reply:
Got it, in case you want to, I think [SequenceEqual](https://docs.microsoft.com/en-us/dotnet/api/system.memoryextensions.sequenceequal?view=net-6.0) can be used

Reply:
The main annoyance with that is that I don't have the string length, so I could either P/Invoke strlen (which isn't necessarily cheap), or write my own routine (in which case I'd just write a string-comparing routine instead of a string-length computation routine. Not a big issue, but this will happen once per assembly, and it's opt-in, so I don't think it's all that important.